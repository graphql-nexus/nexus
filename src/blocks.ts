export { InputDefinitionBlock, OutputDefinitionBlock } from './definitions/definitionBlocks'
export { InterfaceDefinitionBlock } from './definitions/interfaceType'
export { ObjectDefinitionBlock } from './definitions/objectType'
export { UnionDefinitionBlock } from './definitions/unionType'
