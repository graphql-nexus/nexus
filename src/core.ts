// The "core" is used as a namespace to re-export everything,
// For anyone who wants to use the internals
export * from './builder'
export * from './makeSchema'
export * from './definitions/args'
export * from './definitions/decorateType'
export * from './definitions/definitionBlocks'
export * from './definitions/enumType'
export * from './definitions/extendInputType'
export * from './definitions/extendType'
export * from './definitions/inputObjectType'
export * from './definitions/interfaceType'
export * from './definitions/list'
export * from './definitions/mutationField'
export * from './definitions/mutationType'
export * from './definitions/nonNull'
export * from './definitions/nullable'
export * from './definitions/objectType'
export * from './definitions/queryField'
export * from './definitions/queryType'
export * from './definitions/scalarType'
export * from './definitions/subscriptionField'
export * from './definitions/subscriptionType'
export * from './definitions/nexusMeta'
export * from './definitions/unionType'
export * from './definitions/wrapping'
export * from './definitions/_types'
export * from './dynamicMethod'
export * from './plugin'
export * from './plugins'
export * from './rebuildType'
export * from './sdlConverter'
export * from './typegenAutoConfig'
export * from './typegenFormatPrettier'
export * from './typegenMetadata'
export * from './typegenPrinter'
export * from './typegenTypeHelpers'
export * from './typegenUtils'
export * from './utils'
