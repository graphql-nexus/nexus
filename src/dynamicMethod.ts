import { PluginBuilderLens } from './builder'
import { InputDefinitionBlock, OutputDefinitionBlock } from './definitions/definitionBlocks'
import { NexusTypes, withNexusSymbol } from './definitions/_types'

export type OutputFactoryConfig<T> = {
  stage: 'walk' | 'build'
  args: any[]
  builder: PluginBuilderLens
  typeDef: OutputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export type InputFactoryConfig<T> = {
  args: any[]
  builder: PluginBuilderLens
  typeDef: InputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export interface BaseExtensionConfig<T extends string> {
  /**
   * The name of the "extension", the field made
   * available on the builders
   */
  name: T
  /**
   * The full type definition for the options, including generic
   * signature for the type
   */
  typeDefinition?: string
}

export interface DynamicOutputMethodConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the field is called
   */
  factory(config: OutputFactoryConfig<T>): any
}

export interface DynamicInputMethodConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the field is called
   */
  factory(config: InputFactoryConfig<T>): any
}

export class DynamicInputMethodDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicInputMethodConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicInputMethodDef, NexusTypes.DynamicInput)

export class DynamicOutputMethodDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicOutputMethodConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicOutputMethodDef, NexusTypes.DynamicOutputMethod)

/**
 * Defines a new property on the object definition block
 * for an output type, taking arbitrary input to define
 * additional types. See the connectionPlugin:
 *
 * t.connectionField('posts', {
 *   nullable: true,
 *   totalCount(root, args, ctx, info) {
 *     return ctx.user.getTotalPostCount(root.id, args)
 *   },
 *   nodes(root, args, ctx, info) {
 *     return ctx.user.getPosts(root.id, args)
 *   }
 * })
 */
export function dynamicOutputMethod<T extends string>(config: DynamicOutputMethodConfig<T>) {
  return new DynamicOutputMethodDef(config.name, config)
}

/**
 * Same as the outputFieldExtension, but for fields that
 * should be added on as input types.
 */
export function dynamicInputMethod<T extends string>(config: DynamicInputMethodConfig<T>) {
  return new DynamicInputMethodDef(config.name, config)
}
