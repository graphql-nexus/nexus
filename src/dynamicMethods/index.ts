export * from './collection'
export * from './relayConnection'
