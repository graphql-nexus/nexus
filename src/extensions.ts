import { GraphQLNamedType } from 'graphql'
import { DynamicFieldDefs, SchemaConfig } from './builder'
import { NexusOutputFieldConfig } from './definitions/definitionBlocks'
import { NexusInputObjectTypeConfig } from './definitions/inputObjectType'
import { NexusInterfaceTypeConfig } from './definitions/interfaceType'
import { NexusObjectTypeConfig } from './definitions/objectType'
import { RootTypings } from './definitions/_types'

export type NexusGraphQLNamedType = GraphQLNamedType & {
  extensions?: {
    nexus?: {
      config: any
    }
  }
}

export type NexusTypeExtensions = NexusObjectTypeExtension | NexusInterfaceTypeExtension

/**
 * Container object living on `fieldDefinition.extensions.nexus`
 */
export class NexusFieldExtension<TypeName extends string = any, FieldName extends string = any> {
  readonly config: Omit<NexusOutputFieldConfig<TypeName, FieldName>, 'resolve'>
  /**
   * Whether the user has provided a custom "resolve" function,
   * or whether we're using GraphQL's defaultResolver
   */
  readonly hasDefinedResolver: boolean
  constructor(config: NexusOutputFieldConfig<TypeName, FieldName>) {
    const { resolve, ...rest } = config
    this.config = rest
    this.hasDefinedResolver = Boolean(resolve)
  }
}

/**
 * Container object living on `inputObjectType.extensions.nexus`
 */
export class NexusInputObjectTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusInputObjectTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusInputObjectTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

/**
 * Container object living on `objectType.extensions.nexus`
 */
export class NexusObjectTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusObjectTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusObjectTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

/**
 * Container object living on `interfaceType.extensions.nexus`
 */
export class NexusInterfaceTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusInterfaceTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusInterfaceTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

export interface NexusSchemaExtensionConfig extends Omit<SchemaConfig, 'types'> {
  dynamicFields: DynamicFieldDefs
  rootTypings: RootTypings
}

/**
 * Container object living on `schema.extensions.nexus`. Keeps track
 * of metadata from the builder so we can use it when we
 */
export class NexusSchemaExtension {
  constructor(readonly config: NexusSchemaExtensionConfig) {}
}
