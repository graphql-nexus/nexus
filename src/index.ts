import * as blocks from './blocks'
import * as core from './core'

// All of the Public API definitions
export { makeSchema } from './makeSchema'
export { arg, booleanArg, floatArg, idArg, intArg, stringArg } from './definitions/args'
export { decorateType } from './definitions/decorateType'
export { enumType } from './definitions/enumType'
export { extendInputType } from './definitions/extendInputType'
export { extendType } from './definitions/extendType'
export { inputObjectType } from './definitions/inputObjectType'
export { interfaceType } from './definitions/interfaceType'
export { list } from './definitions/list'
export { mutationField } from './definitions/mutationField'
export { mutationType } from './definitions/mutationType'
export { nonNull } from './definitions/nonNull'
export { nullable } from './definitions/nullable'
export { objectType } from './definitions/objectType'
export { queryField } from './definitions/queryField'
export { queryType } from './definitions/queryType'
export { asNexusMethod, scalarType } from './definitions/scalarType'
export { subscriptionField } from './definitions/subscriptionField'
export { subscriptionType } from './definitions/subscriptionType'
export { unionType } from './definitions/unionType'
export { dynamicInputMethod, dynamicOutputMethod } from './dynamicMethod'
export { dynamicOutputProperty } from './dynamicProperty'
export { createPlugin, plugin, PluginBuilderLens, PluginConfig } from './plugin'
export * from './plugins'
export { convertSDL } from './sdlConverter'
export { AllInputTypes, AllOutputTypes, FieldResolver, FieldType } from './typegenTypeHelpers'
export { groupTypes } from './utils'
export { blocks, core }
