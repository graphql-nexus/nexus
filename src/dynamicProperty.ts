import { SchemaBuilder } from './builder'
import { OutputDefinitionBlock } from './definitions/definitionBlocks'
import { NexusTypes, withNexusSymbol } from './definitions/_types'
import { BaseExtensionConfig } from './dynamicMethod'

export type OutputPropertyFactoryConfig<T> = {
  stage: 'walk' | 'build'
  builder: SchemaBuilder
  typeDef: OutputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export interface DynamicOutputPropertyConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the property is accessed (as a getter)
   */
  factory(config: OutputPropertyFactoryConfig<T>): any
}

export class DynamicOutputPropertyDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicOutputPropertyConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicOutputPropertyDef, NexusTypes.DynamicOutputProperty)

/**
 * Defines a new property on the object definition block
 * for an output type, making it possible to build custom DSL's
 * on top of Nexus, e.g. in nexus-prisma
 *
 * t.model.posts()
 */
export function dynamicOutputProperty<T extends string>(config: DynamicOutputPropertyConfig<T>) {
  return new DynamicOutputPropertyDef(config.name, config)
}
