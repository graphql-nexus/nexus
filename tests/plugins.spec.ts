import { lexicographicSortSchema, printSchema } from 'graphql'
import { createPlugin, makeSchema, objectType, PluginConfig, queryType } from '../src'
import { extendType, inputObjectType, NexusAcceptedTypeDef } from '../src/core'

const fooObject = objectType({
  name: 'foo',
  definition(t) {
    t.string('bar')
  },
})

const queryField = extendType({
  type: 'Query',
  definition(t) {
    t.string('something')
  },
})

describe('runtime config validation', () => {
  const spy = jest.spyOn(console, 'error').mockImplementation()

  beforeEach(() => {
    jest.resetAllMocks()
  })

  const whenGiven = (config: any) => () => createPlugin(config)

  it('checks name present', () => {
    expect(whenGiven({})).toThrowErrorMatchingSnapshot()
  })

  it('checks name is string', () => {
    expect(whenGiven({ name: 1 })).toThrowErrorMatchingSnapshot()
  })

  it('checks name is not empty', () => {
    expect(whenGiven({ name: '' })).toThrowErrorMatchingSnapshot()
  })

  it('checks onInstall is a function if defined', () => {
    whenGiven({ name: 'x', onInstall: 'foo' })()
    expect(spy).toBeCalledTimes(1)
    jest.resetAllMocks()
    whenGiven({ name: 'x', onInstall: {} })()
    expect(spy).toBeCalledTimes(1)
  })
})

describe('a plugin may', () => {
  const whenGiven = (pluginConfig: PluginConfig) => () =>
    makeSchema({
      outputs: false,
      types: [],
      plugins: [createPlugin(pluginConfig)],
    })

  it('do nothing', () => {
    expect(whenGiven({ name: 'x' }))
  })
})

describe('onInstall plugins', () => {
  const whenGiven = ({
    onInstall,
    plugin,
    appTypes,
  }: {
    onInstall?: PluginConfig['onInstall']
    plugin?: Omit<PluginConfig, 'name'>
    appTypes?: NexusAcceptedTypeDef[]
  }) => {
    const xPluginConfig = plugin || { onInstall }

    return printSchema(
      lexicographicSortSchema(
        makeSchema({
          outputs: false,
          types: appTypes || [],
          plugins: [createPlugin({ name: 'x', ...xPluginConfig })],
        })
      )
    )
  }

  it('is an optional hook', () => {
    expect(whenGiven({ plugin: {} }))
  })

  it('may return an empty array of types', () => {
    expect(whenGiven({ onInstall: () => {} }))
  })

  it('may contribute types', () => {
    expect(
      whenGiven({
        onInstall: (builder) => {
          builder.addType(queryField)
        },
      })
    ).toMatchSnapshot()
  })

  it('has access to top-level types', () => {
    expect(
      whenGiven({
        onInstall: (builder) => {
          if (!builder.hasType('foo')) {
            builder.addType(queryField)
          }
        },
        appTypes: [fooObject],
      })
    ).toMatchSnapshot()
  })

  it('does not see fallback ok-query', () => {
    expect(
      whenGiven({
        onInstall(builder) {
          if (builder.hasType('Query')) {
            builder.addType(queryField)
          }
        },
      })
    ).toMatchSnapshot()
  })

  it('does not have access to inline types', () => {
    expect(
      whenGiven({
        onInstall: (builder) => {
          if (builder.hasType('Inline')) {
            builder.addType(queryField)
          }
        },
        appTypes: [
          queryType({
            definition(t) {
              t.string('bar', {
                args: {
                  inline: inputObjectType({
                    name: 'Inline',
                    definition(t2) {
                      t2.string('hidden')
                    },
                  }),
                },
              })
            },
          }),
        ],
      })
    ).toMatchSnapshot()
  })
})
