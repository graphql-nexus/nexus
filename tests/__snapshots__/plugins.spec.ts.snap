// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onInstall plugins does not have access to inline types 1`] = `
"input Inline {
  hidden: String
}

type Query {
  bar(inline: Inline): String
}
"
`;

exports[`onInstall plugins does not see fallback ok-query 1`] = `
"type Query {
  ok: Boolean!
}
"
`;

exports[`onInstall plugins has access to top-level types 1`] = `
"type foo {
  bar: String
}

type Query {
  ok: Boolean!
}
"
`;

exports[`onInstall plugins may contribute types 1`] = `
"type Query {
  something: String
}
"
`;

exports[`runtime config validation checks name is not empty 1`] = `"Plugin \\"\\" is giving an invalid value for property name: empty string"`;

exports[`runtime config validation checks name is string 1`] = `"Plugin \\"1\\" is giving an invalid value for property name: expected \\"string\\" type, got number type"`;

exports[`runtime config validation checks name present 1`] = `"Plugin \\"undefined\\" is missing required properties: name"`;
